README
# Flask-Restless-NG #

[![PyPI version](https://badge.fury.io/py/Flask-Restless-NG.svg)](https://badge.fury.io/py/Flask-Restless-NG)
[![Build status](https://secure.travis-ci.org/mrevutskyi/flask-restless-ng.png)](https://travis-ci.org/mrevutskyi/flask-restless-ng)

## About

This is a fork of no longer maintained [Flask-Restless](https://github.com/jfinkels/flask-restless) module.
The goal of this fork is to keep it compatible with the latest version of Flask and SQLAlchemy, performance
improvements (faster serializer, etc) and bug fixing

## Introduction ##

This is Flask-Restless, a [Flask][1] extension that creates URL endpoints that
satisfy the requirements of the [JSON API][2] specification. It is compatible
with models that have been defined using either [SQLAlchemy][3] or
[Flask-SQLAlchemy][4].

This document contains some brief instructions concerning installation of
requirements, installation of this extension, configuration and usage of this
extension, and building of documentation.

For more information, see the

  * [documentation][5],
  * [Python Package Index listing][6],
  * [source code repository][7].

[1]: http://flask.pocoo.org
[2]: https://jsonapi.org
[3]: https://sqlalchemy.org
[4]: https://packages.python.org/Flask-SQLAlchemy
[5]: https://flask-restless.readthedocs.org
[6]: https://pypi.python.org/pypi/Flask-Restless
[7]: https://github.com/jfinkels/flask-restless

## Copyright license ##

The code comprising this program is copyright 2011 Lincoln de Sousa and
copyright 2012, 2013, 2014, 2015, 2016 Jeffrey Finkelstein and contributors,
and is dual-licensed under the following two copyright licenses:

* the GNU Affero General Public License, either version 3 or (at your option)
  any later version
* the 3-clause BSD License

For more information, see the files `LICENSE.AGPL` and `LICENSE.BSD` in this
directory.

The documentation is licensed under the Creative Commons Attribution-ShareAlike
4.0 license.

## Contents ##

This is a partial listing of the contents of this package.

* `LICENSE.AGPL` - one possible copyright license under which this program is
  distributed to you (the GNU Affero General Public License version 3)
* `LICENSE.BSD` - another possible copyright license under which this program
  is distributed to you (the 3-clause BSD License)
* `docs/` - the Sphinx documentation for Flask-Restless
* `examples/` - example applications of Flask-Restless
* `flask_restless/` - the Python package containing the extension
* `MANIFEST.in` - indicates files to include when packaging Flask-Restless
* `README.md` - this file
* `setup.py` - Python setuptools configuration file for packaging this
  extension
* `tests/` - unit tests for Flask-Restless

The `flask_restless` directory is a Python package containing the following
files and directory:

* `helpers.py` - utility functions, mainly for performing introspection on
  SQLAlchemy objects
* `manager.py` - contains the main class that end users will utilize to create
  ReSTful JSON APIs for their database models
* `search.py` - functions and classes that facilitate searching the database
  on requests that require a search
* `serialization.py` - basic serialization and deserialization for SQLAlchemy
  models
* `views/` - the view classes that implement the JSON API interface

## Installing ##-m twine upload --repository testpypi dist/*

This application can be used with any Python version that Flask supports, which
currently includes versions 2.6, 2.7, 3.3, and 3.4. (Python 3.2 is not
supported by Flask and therefore cannot be supported by Flask-Restless.)

This application requires the following libraries to be installed:

* [Flask][1] version 0.10 or greater
* [SQLAlchemy][3] version 0.8 or greater
* [python-dateutil][8] version strictly greater than 2.2
* [mimerender][9] version 0.5.2 or greater

These requirements (and some additional optional packages) are also listed in
the `requirements.txt` file. Using `pip` is probably the easiest way to install
these:

    pip install -r requirements.txt

[8]: https://labix.org/python-dateutil
[9]: https://github.com/martinblech/mimerender

## Building as a Python egg ##

This package can be built, installed, etc. as a Python egg using the provided
`setup.py` script. For more information, run

    python setup.py --help

## How to use ##

For information on how to use this extension, build the documentation here or
[view it on the Web][5].

## Testing ##

Running the unit tests requires the [nosetests][10] library. This requirement
is also listed in the `requirements-test.txt` file.

Using `pip` is probably the easiest way to install this:

    pip install -r requirements-test.txt

To run the tests:

    python -m unittest


## Building documentation ##

Flask-Restless requires the following program and supporting library to build
the documentation:

* [Sphinx][11]
* [sphinxcontrib-httpdomain][12], version 1.1.7 or greater

These requirements are also listed in the `requirements-doc.txt` file. Using
`pip` is probably the easiest way to install these:

    pip install -r requirements-doc.txt

The documentation is written for Sphinx in [reStructuredText][13] files in the
`docs/` directory. Documentation for each class and function is provided in the
docstring in the code.

The documentation uses the Flask Sphinx theme. It is included as a git
submodule of this project, rooted at `docs/_themes`. To get the themes, do

    git submodule update --init

Now to build the documentation, run the command

    python setup.py build_sphinx

in the top-level directory. The output can be viewed in a web browser by
opening `build/sphinx/html/index.html`.

[11]: http://sphinx.pocoo.org/
[12]: https://packages.python.org/sphinxcontrib-httpdomain/
[13]: https://docutils.sourceforge.net/rst.html

## Contributing ##

Please report any issues on the [GitHub Issue Tracker][14].

To suggest a change to the code or documentation, please create a new pull
request on GitHub. Contributed code must come with an appropriate unit
test. Please ensure that your code follows [PEP8][15], by running, for example,
[flake8][16] before submitting a pull request. Also, please squash multiple
commits into a single commit in your pull request by [rebasing][17] onto the
master branch.

By contributing to this project, you are agreeing to license your code
contributions under both the GNU Affero General Public License, either version
3 or any later version, and the 3-clause BSD License, and your documentation
contributions under the Creative Commons Attribution-ShareAlike License version
4.0, as described in the copyright license section above.

[14]: http://github.com/mrevutskyi/flask-restless-ng/issues
[15]: https://www.python.org/dev/peps/pep-0008/
[16]: http://flake8.readthedocs.org/en/latest/
[17]: https://help.github.com/articles/about-git-rebase/

## Artwork ##

The `artwork/flask-restless-small.svg` and
`docs/_static/flask-restless-small.png` are licensed under the [Creative
Commons Attribute-ShareAlike 4.0 license][18]. The original image is a scan of
a (now public domain) illustration by Arthur Hopkins in a serial edition of
"The Return of the Native" by Thomas Hardy published in October 1878.

The `artwork/flask-restless.svg` and `docs/_static/flask-restless.png` are
licensed under the [Flask Artwork License][19].

[18]: https://creativecommons.org/licenses/by-sa/4.0
[19]: http://flask.pocoo.org/docs/license/#flask-artwork-license

## Contact ##

Maksym Revutskyi <maksym.revutskyi@gmail.com>


CHANGES
Changelog
=========

Changes in Flask-Restless-NG
============================

Version 2.0.2
--------------
- Fixed import for SQLAlchemy 1.3 #22


Version 2.0.0
--------------

Refactored fetching resource collections:
- SQL query optimizations for 'include' and 'relationship' objects, using `selectinload`
  (3x-5x performance improvement when tested on large datasets)

- New parameter 'include_links' which controls should relationship objects include links. They are not required by
  JSON API, and disabling them significantly improves performance

- New interfaces for Serializer and Deserializer classes.

- APIManager requires Serializer/Deserializer objects instead of functions for `serializer`/`deserializer` options

Deprecations:
- 'single' parameter is no longer supported - makes code complicated, is not defined in JSON API specs and can be easily
  replicated on a client side

- 'group' parameter is not longer supported - not defined in JSON API specifications, confusing and broken for PostrgeSQL

- JSONP callbacks are no longer supported - please reach out if you have a use case for them


Version 1.0.6
-------------
- Prevent redundant SQL queries during pagination and resource inclusion

Version 1.0.5
-------------
- #16 - Fix: including child of empty relationship (by @sharky98)

Version 1.0.4
-------------
- #15: Support SQLAlchemy 1.4.x

Version 1.0.2
-------------
- #1, #13: Fix for relationship updates not being committed (by @sharky98)
- #12: Fix for 500 when trying to include Null/None relationship
- Added TSQuery operator (by @augustin)

Version 1.0.1
-------------
- #4: `id` is an optional attribute as long as Model has a primary key
- #6: Fix for `flask_restless.views` not being included in the installed package.

Version 1.0.0
-------------
- Performance improvement: url_for() changed to build url locally instead of delegating it to Flask
- This is the last release that is backward compatible with the original Flask-Restless API.


Version 0.0.2
-------------
- New serializer (2-3x faster)
- Added lru_cache to helpers to reduce number of recursive calls (better performance)


Version 0.0.1
-------------

- Fixed 1.0+ compatibility
- Fix for hybrid_property


Original Flask-Restless
=======================
You can find the full `changelog`_ in the original repo

.. _changelog: https://github.com/jfinkels/flask-restless/blob/master/CHANGES


flask_restless/helpers.py
# helpers.py - helper functions for Flask-Restless
#
# Copyright 2011 Lincoln de Sousa <lincoln@comum.org>.
# Copyright 2012, 2013, 2014, 2015, 2016 Jeffrey Finkelstein
#           <jeffrey.finkelstein@gmail.com> and contributors.
#
# This file is part of Flask-Restless.
#
# Flask-Restless is distributed under both the GNU Affero General Public
# License version 3 and under the 3-clause BSD license. For more
# information, see LICENSE.AGPL and LICENSE.BSD.
"""Helper functions for Flask-Restless."""
import datetime
import inspect
from functools import lru_cache
from itertools import chain
from typing import Any
from typing import Dict
from typing import Generator
from typing import Iterable
from typing import List
from typing import Set

from dateutil.parser import parse as parse_datetime
from sqlalchemy import Date
from sqlalchemy import DateTime
from sqlalchemy import Interval
from sqlalchemy import Time
from sqlalchemy.exc import NoInspectionAvailable
from sqlalchemy.ext.hybrid import HYBRID_PROPERTY
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.inspection import inspect as sqlalchemy_inspect
from sqlalchemy.orm import ColumnProperty
from sqlalchemy.orm import Query
from sqlalchemy.orm import RelationshipProperty as RelProperty
from sqlalchemy.orm import class_mapper
from sqlalchemy.orm import load_only
from sqlalchemy.orm import selectinload
from sqlalchemy.orm.attributes import InstrumentedAttribute
from sqlalchemy.orm.attributes import QueryableAttribute
from sqlalchemy.orm.dynamic import DynamicAttributeImpl
from sqlalchemy.sql import func
from sqlalchemy.sql.expression import ColumnElement

try:
    # SQLAlchemy 1.4+
    from sqlalchemy.orm import DeclarativeMeta
except ImportError:
    from sqlalchemy.ext.declarative.api import DeclarativeMeta

try:
    # SQLAlchemy 1.3+
    from sqlalchemy.ext.associationproxy import ObjectAssociationProxyInstance as AssociationProxyType
except ImportError:
    from sqlalchemy.ext.associationproxy import AssociationProxy as AssociationProxyType  # type: ignore

#: Names of attributes which should definitely not be considered relations when
#: dynamically computing a list of relations of a SQLAlchemy model.
RELATION_BLACKLIST = ('query', 'query_class', '_sa_class_manager',
                      '_decl_class_registry')

#: Types which should be considered columns of a model when iterating over all
#: attributes of a model class.
COLUMN_TYPES = (InstrumentedAttribute, hybrid_property)

#: Strings which, when received by the server as the value of a date or time
#: field, indicate that the server should use the current time when setting the
#: value of the field.
CURRENT_TIME_MARKERS = ('CURRENT_TIMESTAMP', 'CURRENT_DATE', 'LOCALTIMESTAMP')


def session_query(session, model):
    """Returns a SQLAlchemy query object for the specified `model`.

    If `model` has a ``query`` attribute already, ``model.query`` will be
    returned. If the ``query`` attribute is callable ``model.query()`` will be
    returned instead.

    If `model` has no such attribute, a query based on `session` will be
    created and returned.

    """
    if hasattr(model, 'query'):
        if callable(model.query):
            query = model.query()
        else:
            query = model.query
        if hasattr(query, 'filter'):
            return query
    return session.query(model)


def get_relations(model):
    """Returns a list of relation names of `model` (as a list of strings)."""
    return [k for k in dir(model)
            if not (k.startswith('_') or k in RELATION_BLACKLIST)
            and get_related_model(model, k)]


@lru_cache()
def get_related_model(model, relationname):
    """Gets the class of the model to which `model` is related by the attribute
    whose name is `relationname`.

    For example, if we have the model classes ::

        class Person(Base):
            __tablename__ = 'person'
            id = Column(Integer, primary_key=True)
            articles = relationship('Article')

        class Article(Base):
            __tablename__ = 'article'
            id = Column(Integer, primary_key=True)
            author_id = Column(Integer, ForeignKey('person.id'))
            author = relationship('Person')

    then

        >>> get_related_model(Person, 'articles')
        <class 'Article'>
        >>> get_related_model(Article, 'author')
        <class 'Person'>

    """
    if hasattr(model, relationname):
        # inspector = sqlalchemy_inspect(model)
        # attributes = inspector.attrs
        # if relationname in attributes:
        #     state = attributes[relationname]
        attr = getattr(model, relationname)
        if hasattr(attr, 'property') \
                and isinstance(attr.property, RelProperty):
            return attr.property.mapper.class_
        if isinstance(attr, AssociationProxyType):
            return get_related_association_proxy_model(attr)
    return None


def get_related_association_proxy_model(attr):
    """Returns the model class specified by the given SQLAlchemy relation
    attribute, or ``None`` if no such class can be inferred.

    `attr` must be a relation attribute corresponding to an association proxy.

    """
    prop = attr.remote_attr.property
    for attribute in ('mapper', 'parent'):
        if hasattr(prop, attribute):
            return getattr(prop, attribute).class_


def foreign_key_columns(model):
    """Returns a list of the :class:`sqlalchemy.Column` objects that contain
    foreign keys for relationships in the specified model class.

    """
    try:
        inspector = sqlalchemy_inspect(model)
    except NoInspectionAvailable:
        # Well, the inspection of a model class returns a mapper anyway, so
        # let's just assume the inspection would have returned the mapper.
        inspector = class_mapper(model)
    all_columns = inspector.columns
    return [c for c in all_columns if c.foreign_keys]


def foreign_keys(model):
    """Returns a list of the names of columns that contain foreign keys for
    relationships in the specified model class.

    """
    return [column.name for column in foreign_key_columns(model)]


def has_field(model, fieldname):
    """Returns ``True`` if the `model` has the specified field or if it has a
    settable hybrid property for this field name.

    """
    descriptors = sqlalchemy_inspect(model).all_orm_descriptors._data
    if fieldname in descriptors and hasattr(descriptors[fieldname], 'fset'):
        return descriptors[fieldname].fset is not None
    return hasattr(model, fieldname)


def get_field_type(model, fieldname):
    """Helper which returns the SQLAlchemy type of the field."""
    field = getattr(model, fieldname)
    if isinstance(field, ColumnElement):
        return field.type
    if isinstance(field, AssociationProxyType):
        field = field.remote_attr
    if hasattr(field, 'property'):
        prop = field.property
        if isinstance(prop, RelProperty):
            return None
        return prop.columns[0].type
    return None


def attribute_columns(model) -> List[str]:
    """Returns a list of model's column names that should be considered as attributes."""
    inspected_model = sqlalchemy_inspect(model)
    column_attrs = inspected_model.column_attrs.keys()
    descriptors = inspected_model.all_orm_descriptors.items()
    hybrid_columns = [k for k, d in descriptors if d.extension_type == HYBRID_PROPERTY]

    return column_attrs + hybrid_columns


@lru_cache()
def primary_key_names(model):
    """Returns all the primary keys for a model."""
    return [key for key, field in inspect.getmembers(model)
            if isinstance(field, QueryableAttribute)
            and hasattr(field, 'property')
            and isinstance(field.property, ColumnProperty)
            and field.property.columns[0].primary_key]


def is_proxy(value: Any) -> bool:
    return isinstance(value, AssociationProxyType)


def is_like_list(instance, relation):
    """Returns ``True`` if and only if the relation of `instance` whose name is
    `relation` is list-like.

    A relation may be like a list if, for example, it is a non-lazy one-to-many
    relation, or it is a dynamically loaded one-to-many.

    """
    if relation in instance._sa_class_manager:
        return instance._sa_class_manager[relation].property.uselist
    elif hasattr(instance, relation):
        attr = getattr(instance._sa_instance_state.class_, relation)
        if hasattr(attr, 'property'):
            return attr.property.uselist
    related_value = getattr(type(instance), relation, None)
    if is_proxy(related_value):
        local_prop = related_value.local_attr.prop
        if isinstance(local_prop, RelProperty):
            return local_prop.uselist
    return False


def query_by_primary_key(session, model, pk_value, primary_key=None):
    """Returns a SQLAlchemy query object containing the result of querying
    `model` for instances whose primary key has the value `pk_value`.

    If `primary_key` is specified, the column specified by that string is used
    as the primary key column. Otherwise, the column named ``id`` is used.

    Presumably, the returned query should have at most one element.

    """
    pk_name = primary_key
    query = session_query(session, model)
    return query.filter(getattr(model, pk_name) == pk_value)


def selectinload_included_relationships(
        model: DeclarativeMeta,
        query: Query,
        include: Set[str],
        relationship_columns: Set[str],
        filters=None
) -> Query:
    join_paths = {path.split('.')[0] for path in include}

    for path in join_paths:
        attribute = getattr(model, path)
        if not is_proxy(attribute) and not isinstance(attribute.impl, DynamicAttributeImpl):
            query = query.options(selectinload(attribute))

    for path in relationship_columns:
        attribute = getattr(model, path)
        if path not in join_paths and not is_proxy(attribute) and not isinstance(attribute.impl, DynamicAttributeImpl):
            options = selectinload(attribute)
            # if request contains filters we need to load all columns
            if not filters:
                options = options.options(load_only('id'))
            query = query.options(options)

    return query


def get_inclusions_for_instances(include: Set[str], instances) -> Set:
    inclusion_tree: Dict[str, dict] = dict()
    for path in include:
        tree = path.split('.')
        current_tree = inclusion_tree
        for level in tree:
            current_tree[level] = current_tree.get(level, {})
            current_tree = current_tree[level]

    return set(chain.from_iterable(get_inclusions(inclusion_tree, instances)))


def get_inclusions(inclusion_tree: Dict[str, dict], instances: Iterable) -> Generator:
    stack = []
    while True:
        for key, sub_tree in inclusion_tree.items():
            new_instances = set()
            for instance in instances:
                included_instance = getattr(instance, key)
                if not included_instance:
                    continue
                if is_like_list(instance, key):
                    new_instances.update(set(included_instance))
                else:
                    new_instances.add(included_instance)
            if sub_tree and new_instances:
                stack.append((sub_tree, new_instances))
            yield new_instances
        if not stack:
            break
        inclusion_tree, instances = stack.pop()


def get_by(session, model, pk_value, primary_key):
    """Returns the first instance of `model` whose primary key has the value
    `pk_value`, or ``None`` if no such instance exists.

    If `primary_key` is specified, the column specified by that string is used
    as the primary key column. Otherwise, the column named ``id`` is used.

    """
    result = query_by_primary_key(session, model, pk_value, primary_key)
    return result.first()


def string_to_datetime(model, fieldname, value):
    """Casts `value` to a :class:`datetime.datetime` or
    :class:`datetime.timedelta` object if the given field of the given
    model is a date-like or interval column.

    If the field name corresponds to a field in the model which is a
    :class:`sqlalchemy.types.Date`, :class:`sqlalchemy.types.DateTime`,
    or :class:`sqlalchemy.Interval`, then the returned value will be the
    :class:`datetime.datetime` or :class:`datetime.timedelta` Python
    object corresponding to `value`. Otherwise, the `value` is returned
    unchanged.

    """
    if value is None:
        return value
    # If this is a date, time or datetime field, parse it and convert it to
    # the appropriate type.
    field_type = get_field_type(model, fieldname)
    if isinstance(field_type, (Date, Time, DateTime)):
        # If the string is empty, no datetime can be inferred from it.
        if value.strip() == '':
            return None
        # If the string is a string indicating that the value of should be the
        # current datetime on the server, get the current datetime that way.
        if value in CURRENT_TIME_MARKERS:
            return getattr(func, value.lower())()
        value_as_datetime = parse_datetime(value)
        # If the attribute on the model needs to be a Date or Time object as
        # opposed to a DateTime object, just get the date component of the
        # datetime.
        if isinstance(field_type, Date):
            return value_as_datetime.date()
        if isinstance(field_type, Time):
            return value_as_datetime.timetz()
        return value_as_datetime
    # If this is an Interval field, convert the integer value to a timedelta.
    if isinstance(field_type, Interval) and isinstance(value, int):
        return datetime.timedelta(seconds=value)
    # In any other case, simply copy the value unchanged.
    return value


def strings_to_datetimes(model, dictionary):
    """Returns a new dictionary with all the mappings of `dictionary` but
    with date strings and intervals mapped to :class:`datetime.datetime` or
    :class:`datetime.timedelta` objects.

    The keys of `dictionary` are names of fields in the model specified in the
    constructor of this class. The values are values to set on these fields. If
    a field name corresponds to a field in the model which is a
    :class:`sqlalchemy.types.Date`, :class:`sqlalchemy.types.DateTime`, or
    :class:`sqlalchemy.Interval`, then the returned dictionary will have the
    corresponding :class:`datetime.datetime` or :class:`datetime.timedelta`
    Python object as the value of that mapping in place of the string.

    This function outputs a new dictionary; it does not modify the argument.

    """
    return {k: string_to_datetime(model, k, v) for k, v in dictionary.items() if k not in ('type', 'links')}


def get_model(instance):
    """Returns the model class of which the specified object is an instance."""
    return type(instance)
